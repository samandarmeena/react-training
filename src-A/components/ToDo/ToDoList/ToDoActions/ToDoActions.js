import React from 'react';
import PropTypes from 'prop-types';
import Image from '../../../Shared/Image';
import Button from '../../../Shared/Button';

/**
 * ToDoActions component
 * renders ToDoActions container
 * @param {string} itemId
 * @param {string} itemStatus
 * @param {func} handleToggleFixed
 * @param {func} handleRemoveToDo
 * @return Element
 */
const ToDoActions = (props) => {
  const { itemId, itemStatus, handleToggleFixed, handleRemoveToDo } = props;
  const actionButtons = [
    {
      buttonName: `complete${itemId}`,
      buttonId: `complete${itemId}`,
      buttonSize: 'buttonSmall',
      buttonToolTip: itemStatus === 'New' ? 'Mark as Completed' : 'Mark as New',
      buttonStyle: itemStatus === 'New' ? 'buttonSuccess' : 'buttonInfo',
      addOnClass: 'changeStatus',
      buttonTitle:
        itemStatus === 'New' ? (
          <Image src="/assets/images/check.png" alt="Check" />
        ) : (
          <Image src="/assets/images/return.png" alt="Return" />
        ),
      clickEvent: () => handleToggleFixed(itemId)
    },
    {
      buttonName: `remove${itemId}`,
      buttonId: `remove${itemId}`,
      buttonSize: 'buttonSmall',
      buttonStyle: 'buttonDanger',
      addOnClass: 'removeToDo',
      buttonToolTip: 'Remove ToDo',
      buttonTitle: <Image src="/assets/images/trash.png" alt="Remove" />,
      clickEvent: () => handleRemoveToDo(itemId)
    }
  ];
  return (
    <>
      {actionButtons.map((actionButton) => {
        return (
          <Button
            key={actionButton.buttonId}
            buttonName={actionButton.buttonName}
            buttonId={actionButton.buttonId}
            buttonSize={actionButton.buttonSize}
            buttonStyle={actionButton.buttonStyle}
            addOnClass={actionButton.addOnClass}
            buttonToolTip={actionButton.buttonToolTip}
            buttonTitle={actionButton.buttonTitle}
            clickEvent={actionButton.clickEvent}
          />
        );
      })}
    </>
  );
};

// proptype validations for props
ToDoActions.propTypes = {
  itemId: PropTypes.number.isRequired,
  itemStatus: PropTypes.string.isRequired,
  handleToggleFixed: PropTypes.func.isRequired,
  handleRemoveToDo: PropTypes.func.isRequired
};

export default ToDoActions;
