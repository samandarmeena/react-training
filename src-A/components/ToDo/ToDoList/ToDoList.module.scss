.toDoList {
  padding: 0;
  .listItems {
    display: flex;
    font-size: 15px;
    border: 1px solid #ccc;
    margin-bottom: 10px;
    padding: 5px 10px;
    border-radius: 5px;
    align-items: center;
    justify-content: space-between;
    .toDoTitle {
      font-size: 15px;
    }
    .toDoActions {
      float: right;
    }
  }
}
