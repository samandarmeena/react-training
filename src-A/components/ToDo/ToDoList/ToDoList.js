import React from 'react';
import PropTypes from 'prop-types';
import ToDoActions from './ToDoActions';
import styles from './ToDoList.module.scss';

/**
 * ToDoList component
 * renders ToDoList container
 * @param {array} todoItems
 * @param {string} filterBy
 * @param {func} toggleFixed
 * @param {func} removeToDo
 * @return Element
 */
const ToDoList = (props) => {
  const { todoItems, filterBy, toggleFixed, removeToDo } = props;
  return (
    <ul className={styles.toDoList}>
      {todoItems.map((todoItem) => {
        return todoItem.status === filterBy ? (
          <li key={todoItem.id} className={styles.listItems}>
            <span className={styles.toDoTitle}>{todoItem.title}</span>
            <span className={styles.toDoActions}>
              <ToDoActions
                itemId={todoItem.id}
                itemStatus={todoItem.status}
                handleToggleFixed={toggleFixed}
                handleRemoveToDo={removeToDo}
              />
            </span>
          </li>
        ) : (
          ''
        );
      })}
    </ul>
  );
};

// proptype validations for props
ToDoList.propTypes = {
  todoItems: PropTypes.array.isRequired,
  filterBy: PropTypes.string.isRequired,
  toggleFixed: PropTypes.func.isRequired,
  removeToDo: PropTypes.func.isRequired
};

export default ToDoList;
