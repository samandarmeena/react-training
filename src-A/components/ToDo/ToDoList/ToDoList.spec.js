import React from 'react';
import { mount } from 'enzyme';
import ToDoList from './ToDoList';
import ToDoActions from './ToDoActions';

describe('ToDoList component', () => {
  let wrapper = '';
  let inputProps = {};
  const toggleFixedFunction = jest.fn();
  const removeToDoFunction = jest.fn();
  beforeAll(() => {
    inputProps = {
      todoItems: [],
      filterBy: 'New',
      toggleFixed: toggleFixedFunction,
      removeToDo: removeToDoFunction
    };
    wrapper = mount(<ToDoList {...inputProps} />);
  });

  it('ToDoList component rendered without an error', () => {
    expect(wrapper).toBeTruthy();
  });

  it('Check length of the children', () => {
    expect(wrapper).toHaveLength(1);
  });

  it('Check length of the list', () => {
    expect(wrapper.find('ul').length).toBe(1);
  });

  it('Default list items should be zero', () => {
    expect(wrapper.find('ul').find('li').length).toBe(0);
  });

  it('should render 2 items in the list', () => {
    let newinputProps = {
      ...inputProps,
      todoItems: [
        { id: 1, title: 'Test todo title 1', status: 'New' },
        { id: 2, title: 'Test todo title 2', status: 'New' }
      ]
    };
    wrapper = mount(<ToDoList {...newinputProps} />);
    expect(wrapper.find('ul').find('li').length).toBe(2);
  });

  it('should render ToDoActions children in each list item', () => {
    let newinputProps = {
      ...inputProps,
      todoItems: [
        { id: 1, title: 'Test todo title 1', status: 'New' },
        { id: 2, title: 'Test todo title 2', status: 'New' }
      ]
    };
    wrapper = mount(<ToDoList {...newinputProps} />);
    expect(
      wrapper
        .find('ul')
        .find('li')
        .first()
        .find(ToDoActions).length
    ).toBe(1);
  });

  it('Snapshot testing', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
