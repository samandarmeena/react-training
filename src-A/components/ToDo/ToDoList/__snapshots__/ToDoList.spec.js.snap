// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoList component Snapshot testing 1`] = `
<ToDoList
  filterBy="New"
  removeToDo={[MockFunction]}
  todoItems={
    Array [
      Object {
        "id": 1,
        "status": "New",
        "title": "Test todo title 1",
      },
      Object {
        "id": 2,
        "status": "New",
        "title": "Test todo title 2",
      },
    ]
  }
  toggleFixed={[MockFunction]}
>
  <ul
    className="toDoList"
  >
    <li
      className="listItems"
      key="1"
    >
      <span
        className="toDoTitle"
      >
        Test todo title 1
      </span>
      <span
        className="toDoActions"
      >
        <ToDoActions
          handleRemoveToDo={[MockFunction]}
          handleToggleFixed={[MockFunction]}
          itemId={1}
          itemStatus="New"
        >
          <Button
            addOnClass="changeStatus"
            buttonId="complete1"
            buttonName="complete1"
            buttonSize="buttonSmall"
            buttonStyle="buttonSuccess"
            buttonTitle={
              <Image
                alt="Check"
                src="/assets/images/check.png"
              />
            }
            buttonToolTip="Mark as Completed"
            buttonType="button"
            clickEvent={[Function]}
            key="complete1"
          >
            <button
              className="button buttonSuccess buttonSmall changeStatus"
              id="complete1"
              name="complete1"
              onClick={[Function]}
              title="Mark as Completed"
              type="button"
            >
              <Image
                alt="Check"
                src="/assets/images/check.png"
              >
                <img
                  alt="Check"
                  src="/assets/images/check.png"
                />
              </Image>
            </button>
          </Button>
          <Button
            addOnClass="removeToDo"
            buttonId="remove1"
            buttonName="remove1"
            buttonSize="buttonSmall"
            buttonStyle="buttonDanger"
            buttonTitle={
              <Image
                alt="Remove"
                src="/assets/images/trash.png"
              />
            }
            buttonToolTip="Remove ToDo"
            buttonType="button"
            clickEvent={[Function]}
            key="remove1"
          >
            <button
              className="button buttonDanger buttonSmall removeToDo"
              id="remove1"
              name="remove1"
              onClick={[Function]}
              title="Remove ToDo"
              type="button"
            >
              <Image
                alt="Remove"
                src="/assets/images/trash.png"
              >
                <img
                  alt="Remove"
                  src="/assets/images/trash.png"
                />
              </Image>
            </button>
          </Button>
        </ToDoActions>
      </span>
    </li>
    <li
      className="listItems"
      key="2"
    >
      <span
        className="toDoTitle"
      >
        Test todo title 2
      </span>
      <span
        className="toDoActions"
      >
        <ToDoActions
          handleRemoveToDo={[MockFunction]}
          handleToggleFixed={[MockFunction]}
          itemId={2}
          itemStatus="New"
        >
          <Button
            addOnClass="changeStatus"
            buttonId="complete2"
            buttonName="complete2"
            buttonSize="buttonSmall"
            buttonStyle="buttonSuccess"
            buttonTitle={
              <Image
                alt="Check"
                src="/assets/images/check.png"
              />
            }
            buttonToolTip="Mark as Completed"
            buttonType="button"
            clickEvent={[Function]}
            key="complete2"
          >
            <button
              className="button buttonSuccess buttonSmall changeStatus"
              id="complete2"
              name="complete2"
              onClick={[Function]}
              title="Mark as Completed"
              type="button"
            >
              <Image
                alt="Check"
                src="/assets/images/check.png"
              >
                <img
                  alt="Check"
                  src="/assets/images/check.png"
                />
              </Image>
            </button>
          </Button>
          <Button
            addOnClass="removeToDo"
            buttonId="remove2"
            buttonName="remove2"
            buttonSize="buttonSmall"
            buttonStyle="buttonDanger"
            buttonTitle={
              <Image
                alt="Remove"
                src="/assets/images/trash.png"
              />
            }
            buttonToolTip="Remove ToDo"
            buttonType="button"
            clickEvent={[Function]}
            key="remove2"
          >
            <button
              className="button buttonDanger buttonSmall removeToDo"
              id="remove2"
              name="remove2"
              onClick={[Function]}
              title="Remove ToDo"
              type="button"
            >
              <Image
                alt="Remove"
                src="/assets/images/trash.png"
              >
                <img
                  alt="Remove"
                  src="/assets/images/trash.png"
                />
              </Image>
            </button>
          </Button>
        </ToDoActions>
      </span>
    </li>
  </ul>
</ToDoList>
`;
